src/uucore/src/lib/features/process.rs

```diff
@@ -100,3 +121,470 @@ impl ChildExt for Child {
+        self.cached_status = Some(Rc::clone(&result));
+        Rc::clone(&result)
+    }
+
+    /// Collect information from `/proc/<pid>/stat` file
+    fn stat(&mut self) -> Rc<Vec<String>> {
+        if let Some(c) = &self.cached_stat {
+            return Rc::clone(c);
+        }
+
+        let result: Vec<_> = stat_split(&self.inner_stat);
+
+        let result = Rc::new(result);
+        self.cached_stat = Some(Rc::clone(&result));
+        Rc::clone(&result)
+    }
+
+    /// Fetch start time
+    ///
+    /// - [The /proc Filesystem: Table 1-4](https://docs.kernel.org/filesystems/proc.html#id10)
+    pub fn start_time(&mut self) -> Result<u64, io::Error> {
+        if let Some(time) = self.cached_start_time {
+            return Ok(time);
+        }
+
+        // Kernel doc: https://docs.kernel.org/filesystems/proc.html#process-specific-subdirectories
+        // Table 1-4
+        let time = self
+            .stat()
+            .get(21)
+            .ok_or(io::ErrorKind::InvalidData)?
+            .parse::<u64>()
+            .map_err(|_| io::ErrorKind::InvalidData)?;
+
+        self.cached_start_time = Some(time);
+
+        Ok(time)
+    }
+
+    /// Fetch run state
+    ///
+    /// - [The /proc Filesystem: Table 1-4](https://docs.kernel.org/filesystems/proc.html#id10)
+    ///
+    /// # Error
+    ///
+    /// If parsing failed, this function will return [io::ErrorKind::InvalidInput]
+    pub fn run_state(&mut self) -> Result<RunState, io::Error> {
+        RunState::try_from(self.stat().get(2).unwrap().as_str())
+    }
+
+    /// This function will scan the `/proc/<pid>/fd` directory
+    ///
+    /// # Error
+    ///
+    /// If scanned pid had mismatched permission,
+    /// it will caused [std::io::ErrorKind::PermissionDenied] error.
+    pub fn ttys(&mut self) -> Result<Rc<HashSet<TerminalType>>, io::Error> {
+        if let Some(tty) = &self.cached_tty {
+            return Ok(Rc::clone(tty));
+        }
+
+        let path = PathBuf::from(format!("/proc/{}/fd", self.pid));
+
+        let result = Rc::new(
+            fs::read_dir(path)?
+                .flatten()
+                .filter(|it| it.path().is_symlink())
+                .flat_map(|it| fs::read_link(it.path()))
+                .flat_map(TerminalType::try_from)
+                .collect::<HashSet<_>>(),
+        );
+
+        self.cached_tty = Some(Rc::clone(&result));
+
+        Ok(result)
+    }
+}
+
+impl TryFrom<DirEntry> for ProcessInformation {
+    type Error = io::Error;
+
+    fn try_from(value: DirEntry) -> Result<Self, Self::Error> {
+        let value = value.into_path();
+
+        Self::try_new(value)
+    }
+}
+
+/// Parsing `/proc/self/stat` file.
+///
+/// In some case, the first pair (and the only one pair) will contains whitespace,
+/// so if we want to parse it, we have to write new algorithm.
+///
+/// TODO: If possible, test and use regex to replace this algorithm.
+fn stat_split(stat: &str) -> Vec<String> {
```

Comment:
did you consider using string slices here ?

Reply:
I think using Vec is sufficient for the needs.

But I think I'll give it a try.

Reply:
Nah I think it's a bit difficult for me now, so sorry :(